File: 001-two-sum/solution.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        i = 0
        d = dict()
        for num in nums:
            d[num] = i
            i = i + 1

        i = 0
        for num in nums:
            if d.has_key(target-num) and d[target-num] != i:
                return [i, d[target-num]]
            i = i + 1
        return [-1, -1]

    # solution2
    def twoSum1(self, nums, target):
        if len(nums) <= 1:
            return [-1, -1]
        d = {}
        for i in range(len(nums)):
            if nums[i] in d:
                return [d[nums[i]], i]
            else:
                d[target-nums[i]] = i
        return [-1, -1]

File: 001-two-sum/Solution.java
public class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        // 将值作为key，索引作为value
        for (int i = 0; i < nums.length; ++i)
            map.put(nums[i], i);

        for (int i = 0; i < nums.length; ++i) {
            // 题目中有个条件：you may not use the same element twice
            if (map.containsKey(target - nums[i]) && (map.get(target - nums[i]) != i))
                return new int[]{i, map.get(target - nums[i])};
        }

        return new int[]{-1, -1};
    }
}